File: android/app/src/main/kotlin/com/rechain/dapp/CrashReportingManager.kt
package com.rechain.dapp

import android.content.Context
import com.google.firebase.crashlytics.FirebaseCrashlytics
import timber.log.Timber

class CrashReportingManager private constructor() {
    
    companion object {
        @Volatile
        private var INSTANCE: CrashReportingManager? = null
        
        fun getInstance(): CrashReportingManager {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: CrashReportingManager().also { INSTANCE = it }
            }
        }
    }
    
    private var crashlytics: FirebaseCrashlytics? = null
    private var isInitialized = false
    
    fun initialize(context: Context, enableInDebug: Boolean = false) {
        if (isInitialized) return
        
        try {
            crashlytics = FirebaseCrashlytics.getInstance()
            
            // Enable/disable crashlytics collection
            crashlytics?.setCrashlyticsCollectionEnabled(
                !BuildConfig.DEBUG || enableInDebug
            )
            
            isInitialized = true
            Timber.d("CrashReportingManager initialized")
            
        } catch (e: Exception) {
            Timber.e(e, "Failed to initialize crash reporting")
        }
    }
    
    fun logException(throwable: Throwable) {
        try {
            crashlytics?.recordException(throwable)
            Timber.e(throwable, "Exception logged to crash reporting")
        } catch (e: Exception) {
            Timber.e(e, "Failed to log exception to crash reporting")
        }
    }
    
    fun logMessage(message: String) {
        try {
            crashlytics?.log(message)
            Timber.d("Message logged to crash reporting: $message")
        } catch (e: Exception) {
            Timber.e(e, "Failed to log message to crash reporting")
        }
    }
    
    fun setUserId(userId: String) {
        try {
            crashlytics?.setUserId(userId)
            Timber.d("User ID set in crash reporting: $userId")
        } catch (e: Exception) {
            Timber.e(e, "Failed to set user ID in crash reporting")
        }
    }
    
    fun setCustomKey(key: String, value: String) {
        try {
            crashlytics?.setCustomKey(key, value)
            Timber.d("Custom key set in crash reporting: $key = $value")
        } catch (e: Exception) {
            Timber.e(e, "Failed to set custom key in crash reporting")
        }
    }
    
    fun setCustomKey(key: String, value: Boolean) {
        try {
            crashlytics?.setCustomKey(key, value)
            Timber.d("Custom key set in crash reporting: $key = $value")
        } catch (e: Exception) {
            Timber.e(e, "Failed to set custom key in crash reporting")
        }
    }
    
    fun setCustomKey(key: String, value: Int) {
        try {
            crashlytics?.setCustomKey(key, value)
            Timber.d("Custom key set in crash reporting: $key = $value")
        } catch (e: Exception) {
            Timber.e(e, "Failed to set custom key in crash reporting")
        }
    }
}

File: macos/Runner/CrashReportingManager.swift
import Foundation
import FirebaseCrashlytics
import os.log

class CrashReportingManager {
    static let shared = CrashReportingManager()
    private let logger = OSLog(subsystem: "com.rechain.online", category: "CrashReporting")
    private var isInitialized = false
    
    private init() {}
    
    func initialize(enableInDebug: Bool = false) {
        #if DEBUG
        guard enableInDebug else {
            os_log("Crash reporting disabled in debug mode", log: logger, type: .info)
            return
        }
        #endif
        
        guard !isInitialized else {
            os_log("Crash reporting already initialized", log: logger, type: .warning)
            return
        }
        
        // Firebase Crashlytics is automatically initialized with Firebase
        isInitialized = true
        os_log("Crash reporting initialized", log: logger, type: .info)
        
        // Set up custom crash handler
        setupCustomCrashHandler()
    }
    
    private func setupCustomCrashHandler() {
        NSSetUncaughtExceptionHandler { exception in
            CrashReportingManager.shared.recordException(exception)
        }
    }
    
    func recordError(_ error: Error, userInfo: [String: Any]? = nil) {
        guard isInitialized else { return }
        
        let nsError = error as NSError
        Crashlytics.crashlytics().record(error: nsError, userInfo: userInfo)
        
        os_log("Error recorded: %@ - %@", log: logger, type: .error, 
               nsError.domain, nsError.localizedDescription)
    }
    
    func recordException(_ exception: NSException) {
        guard isInitialized else { return }
        
        Crashlytics.crashlytics().record(exceptionModel: ExceptionModel(
            name: exception.name.rawValue,
            reason: exception.reason ?? "Unknown reason"
        ))
        
        os_log("Exception recorded: %@ - %@", log: logger, type: .error,
               exception.name.rawValue, exception.reason ?? "Unknown")
    }
    
    func log(_ message: String, level: OSLogType = .info) {
        guard isInitialized else { return }
        
        Crashlytics.crashlytics().log(message)
        os_log("%@", log: logger, type: level, message)
    }
    
    func setUserId(_ userId: String) {
        guard isInitialized else { return }
        
        Crashlytics.crashlytics().setUserID(userId)
        os_log("User ID set: %@", log: logger, type: .info, userId)
    }
    
    func setCustomKey(_ key: String, value: Any) {
        guard isInitialized else { return }
        
        Crashlytics.crashlytics().setCustomValue(value, forKey: key)
        os_log("Custom key set: %@ = %@", log: logger, type: .debug, key, String(describing: value))
    }
    
    func setCustomKeys(_ keys: [String: Any]) {
        guard isInitialized else { return }
        
        for (key, value) in keys {
            setCustomKey(key, value: value)
        }
    }
    
    func recordNonFatalError(_ error: Error, context: [String: Any]? = nil) {
        guard isInitialized else { return }
        
        var userInfo: [String: Any] = [:]
        if let context = context {
            userInfo.merge(context) { _, new in new }
        }
        
        let nsError = error as NSError
        Crashlytics.crashlytics().record(error: nsError, userInfo: userInfo)
        
        os_log("Non-fatal error recorded: %@", log: logger, type: .error, nsError.localizedDescription)
    }
    
    func sendUnsentReports() {
        guard isInitialized else { return }
        
        Crashlytics.crashlytics().sendUnsentReports()
        os_log("Unsent reports sent", log: logger, type: .info)
    }
    
    func deleteUnsentReports() {
        guard isInitialized else { return }
        
        Crashlytics.crashlytics().deleteUnsentReports()
        os_log("Unsent reports deleted", log: logger, type: .info)
    }
    
    func didCrashOnPreviousExecution() -> Bool {
        guard isInitialized else { return false }
        
        return Crashlytics.crashlytics().didCrashDuringPreviousExecution()
    }
}
